ARM64 framepointers

tmcgilchrist opened the pull request:
# ARM64 frame pointers

Frame pointers are used to walk the stack of function calls in a program. This stack-walking technique is commonly used by external profilers and debuggers, including Linux perf and eBPF, FreeBSD and macOS Dtrace, and GDB and LLDB, to reconstruct the call graph for programs. Enabling frame pointers opens up opportunities for third party debugging and performance tools to understand OCaml programs, without needing OCaml specific changes to understand OCaml's calling conventions.

The OCaml compiler already has an option for `enable-frame-pointers` since 4.14 on amd64.
In OCaml 5.1 support was added back for Linux [#11144: Restore frame-pointers support for amd64](https://github.com/ocaml/ocaml/pull/11144) with macOS [#13163](https://github.com/ocaml/ocaml/pull/13163) support coming in 5.3.

This PR extends frame pointer support to ARM64 macOS and Linux.

## Changes

b21e2e6 Stores sp in temporary register x19 during Iextcall in the same way as amd64 backend.
Previously this was implemented  https://github.com/ocaml/ocaml/pull/13079 as push/pop on the stack which
interferes with frame pointers.

2f23134 Implements frame pointer save and restore in both the code generation and arm64 runtime. Using the `stp` and `ldp` instructions adds an extra `add` plus saving an extra register:

``` assembly
;; Lprologue generates
	stp	x29, x30, [sp, #-16]
	sub	sp, sp, #16
	add	x29,  sp, #0

;; output_epilogue
	add	sp, sp, #16
	ldp	x29, x30, [sp, #-16]
```

We already maintain the frame pointer (x29) register when calling into C, plus the [ABI on macOS requires it](https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms) and supporting Linux couldn't hurt. There is minimal extra stack space usage, as we needed to be quad word aligned anyway. I borrowed the same `UPDATE_BASE_POINTER` and `Lentertrap` tricks from https://github.com/ocaml/ocaml/pull/11144 and https://github.com/ocaml/ocaml/pull/11031.

## Performance

For macOS there is no noticeable performance difference so far. I put this down to macOS already having frame pointers turned on. Further testing is needed.

I am still running testing on Linux ARM64. Note that Ubuntu 24.04 LTS [turned on frame pointers by default](https://ubuntu.com/blog/ubuntu-performance-engineering-with-frame-pointers-by-default).

dra27 committed:
Include CAML_LD_LIBRARY_PATH in ocamlrun -config

dra27 committed:
Improve linking error for force-custom libraries

When shared libraries are disabled, .cma files with stubs set a flag to
force -custom mode. -use-prims and -use-runtime conflict -custom, but
the error message was very confusing (since the linker attempts to load
a shared object which cannot have been built).

Alter the linker to detect this case and display a specific error
message.

dra27 committed:
Remove stdlib/hashbang

It was briefly used (as stdlib/sharpbang) during the development of
OCaml 2.00, but replaced with a direct rule before its release.

dra27 committed:
Slightly more "standard" setting of OCAMLRUNPARAM

dra27 committed:
Ensure runtime events were actually dropped

tmcgilchrist committed:
Allow configure --enable-frame-pointers for ARM64 Linux and macOS

MisterDA committed:
Annotate alloc/free open/close functions with compiler attributes

This helps the compiler optimize code, and static analysis by
detecting potential mismatches in alloc/free pairs.

- malloc

> The malloc attribute indicates that the function acts like a system
> memory allocation function, returning a pointer to allocated storage
> disjoint from the storage for any other object accessible to the
> caller.

https://clang.llvm.org/docs/AttributeReference.html#malloc

> Associating a function with a deallocator helps detect calls to
> mismatched allocation and deallocation functions and diagnose them
> under the control of options such as -Wmismatched-dealloc. It also
> makes it possible to diagnose attempts to deallocate objects that
> were not allocated dynamically, by -Wfree-nonheap-object. To
> indicate that an allocation function both satisifies the nonaliasing
> property and has a deallocator associated with it, both the plain
> form of the attribute and the one with the deallocator argument must
> be used.

> The warnings guarded by -fanalyzer respect allocation and
> deallocation pairs marked with the malloc.

https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-malloc-function-attribute

Note that the malloc attribute can only be applied to functions
returning pointers. The OCaml value type is a typedef to an integer
type, and the C compiler will refuse applying the attribute to a
function returning an OCaml value.

- nodiscard / warn_unused_result

Prevent memory leaks by warning if the result of an allocation is
ignored.

https://clang.llvm.org/docs/AttributeReference.html#nodiscard-warn-unused-result

- alloc_align

> GCC uses this information to improve pointer alignment analysis.

https://clang.llvm.org/docs/AttributeReference.html#alloc-align
https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-alloc_005falign-function-attribute

- alloc_size

> GCC uses this information to improve the results of
> __builtin_object_size.

https://clang.llvm.org/docs/AttributeReference.html#alloc-size
https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-alloc_005fsize-function-attribute

- returns_nonnull

> lets the compiler optimize callers based on the knowledge that the
> return value will never be null.

> The analyzer considers the possibility that an allocation function
> could fail and return null. […] If the allocator always returns
> non-null, use __attribute__ ((returns_nonnull)) to suppress these
> warnings.

https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-returns_005fnonnull-function-attribute

tmcgilchrist committed:
Store sp in temporary register x19 during Iextcall

tmcgilchrist committed:
Add frame pointers support for ARM64 Linux and macOS.

gasche committed:
[refactor] ocamltest: refactor run_test_tree's constant parameters

gasche committed:
[refactor] ocamltest: refactor run_test_tree

gasche committed:
[refactor] ocamltest: refactor run_test_tree

nojb committed:
Type-directed array literal disambiguation

nojb committed:
is_principal, expand_head

nojb committed:
Add test

nojb committed:
Remove duplicated function is_principal

nojb committed:
Simplify

nojb committed:
Share code

nojb committed:
Call instance on Predef.type_float

nojb committed:
Share code

yallop committed:
Add support for top-level floatarray printing. (#13)

nojb committed:
Warn if type disambiguation is not principal

tmcgilchrist committed:
Add forty_two test from PR#11031 as exceptions.ml

With frame-pointers, exception handlers should restore RBP #11031

tmcgilchrist committed:
Work around clang/gcc not inlining C functions in the runtime

dustanddreams commented on asmcomp/arm64/emit.mlp:
```diff
@@ -954,7 +969,10 @@ let emit_instr env i =
         let n = frame_size env in
         if env.f.fun_frame_required then
-          `	ldr	{emit_reg i.res.(0)}, [sp, #{emit_int (n-8)}]\n`
+          if fp then
+            `	ldr	{emit_reg i.res.(0)}, [sp, #{emit_int (n-16)}]\n`
+          else
+            `	ldr	{emit_reg i.res.(0)}, [sp, #{emit_int (n-8)}]\n`
```
That chunk is incorrect - `Ireturn_addr` loads the return address saved on stack. Or arm64, this is `x30`, and it is always saved at `[sp + frame size - 8]`.

dustanddreams commented on runtime/arm64.S:
```diff
@@ -1029,6 +1055,7 @@ L(do_perform):
         ldr     x9, Caml_state(current_stack)
         SWITCH_OCAML_STACKS x9, x10
     /*  No parent stack. Raise Effect.Unhandled. */
+        ENTER_FUNCTION
```
This looks incorrect to me - after the TSAN block, this ends with
`b G(caml_c_call)`, which does the `ENTER_FUNCTION`.

amd64 needs `ENTER_FUNCTION` here because it jumps not in the beginning of `caml_c_call` but a bit later down, skipping the `ENTER_FUNCTION`, and compensates here.

tmcgilchrist replied:
Understood, I've removed that part as it was clearly wrong.

dustanddreams commented on runtime/arm64.S:
```diff
@@ -1104,10 +1134,15 @@ FUNCTION(caml_resume)
         mov     x0, x2
         br      x4
-1:      ADDRGLOBAL(ADDITIONAL_ARG, caml_raise_continuation_already_resumed)
+1:      ENTER_FUNCTION
```
Same comment as above - `ENTER_FUNCTION` not needed here on arm64.

tmcgilchrist replied:
Same here.

nojb committed:
Merge remote-tracking branch 'upstream/trunk' into type_directed_array_literals

nojb committed:
Add doc to manual 'Language extensions'

nojb committed:
Changes

nojb committed:
check-typo

fabbing commented on testsuite/tests/frame-pointers/forty_two.ml:
```diff
@@ -0,0 +1,32 @@
+
+let [@inline never] catcher i =
+  Printf.printf "# catcher %d\n%!" i;
+  fp_backtrace Sys.argv.(0);
```
It looks like we forgot to remove the need for `Sys.argv.(0)` aka `execname` when we changed the regular expression filtering the backtrace.

tmcgilchrist replied:
Good catch I'll append that change to this PR when I open it the main ocaml repository.

jmid committed:
Remove unused OCAMLRUNPARAMs list from manual

nojb committed:
Review

tmcgilchrist commented:
Using this test program compiled with `ocamlopt sched.ml -o sched.exe` in a switch with this change on Ubuntu 24.04 LTS (which is compiled with frame pointers turned on):
```ocaml
open Effect
open Effect.Deep

exception E
type _ t += Yield : unit t
          | Fork : (unit -> string) -> unit t
          | Ping : unit t
exception Pong

let say _ =
  (* Printf.printf "%s%!" x *)
  ignore(Sys.opaque_identity(Array.make 1_000 0))

let run main =
  let run_q = Queue.create () in
  let enqueue k = Queue.push k run_q in
  let rec dequeue () =
    if Queue.is_empty run_q then `Finished
    else continue (Queue.pop run_q) ()
  in
  let rec spawn f =
    let[@inline never] spawn_retc = function
      | "ok" -> say "."; dequeue ()
      | s -> failwith ("Unexpected result: " ^ s)
    in
    let[@inline never] spawn_exnc = function
      | E -> say "!"; dequeue ()
      | e -> raise e
    in
    let[@inline never] spawn_effc (type a) (e : a t) =
      match e with
        | Yield -> Some (fun (k : (a, _) continuation) ->
            say ","; enqueue k; dequeue ())
        | Fork f -> Some (fun (k : (a, _) continuation) ->
            say "+"; enqueue k; spawn f)
        | Ping -> Some (fun (k : (a, _) continuation) ->
            say "["; discontinue k Pong)
        | _ -> None
    in
    match_with f ()
    { retc = spawn_retc;
      exnc = spawn_exnc;
      effc = spawn_effc }
  in
  spawn main

let[@inline never] g () =
  let[@inline never] g_comp () = perform Ping; failwith "no pong" in
  let[@inline never] g_retc x = x in
  let[@inline never] g_exnc = function | Pong -> say "]" | e -> raise e in
  let[@inline never] g_effc (type a) (e : a t) =
    match e with
    | Yield -> Some (fun (k : (a,_) continuation) -> failwith "what?")
    | _ -> None in
  perform Yield; say "C"; perform Yield;
  begin match_with g_comp ()
  { retc = g_retc;
    exnc = g_exnc;
    effc = g_effc }
  end;
  raise E

let[@inline never] h () = say "B"; "ok"

let[@inline never] test () =
  say "A";
  perform (Fork g);
  perform (Fork h);
  say "D";
  perform Yield;
  say "E";
  "ok"

let () =
  for a = 1 to 100_000 do
  let `Finished = run test in
  say "\n"
  done
```
First using `sudo perf record --call-graph dwarf -o sched-perf-dwarf.data ./sched.exe`, note `cfree` pointing to an address.
<img width="934" alt="Screenshot 2024-09-25 at 2 57 34 PM" src="https://github.com/user-attachments/assets/70084d4d-22bb-4b21-a28f-57c8042dde9b">

Now with `sudo perf record --call-graph fp -o sched-perf-fp.data ./sched.exe` the `cfree` node is correctly included.
<img width="933" alt="Screenshot 2024-09-25 at 2 57 17 PM" src="https://github.com/user-attachments/assets/040e565d-cb3b-47c0-afb6-1638b28dca98">

nmote committed:
Fix small grammatical error in Changes

See #13203

shym committed:
Escape `'\r'` in `sak`

Add the missing `'\r'` to the characters that must be escaped in all
cases in `sak`

dra27 committed:
Restore the caravan on AppVeyor 🐫 (#13073)

* Revert "Work-around Cygwin 3.5.x regression"

This reverts commit 6b764e6fcd19856ec2bd9fdb41ce0edbac053bf2.

* Allow manually forcing AppVeyor Cygwin upgrade

shindere committed:
Fix syntax error in configure.ac

Remove spurious commas which were causing syntax errors e.g. on
arm32.

dra27 committed:
Merge pull request #13480 from shindere/fix-12019

Fix build on arm32

Octachron committed:
Merge pull request #13478 from nmote/patch-1

Fix small grammatical error in Changes

Commit:
Make a cryptic comment slightly more helpful.

shindere committed:
ocamltest should report skipped test as skipped rather than passed

shindere committed:
Merge pull request #13483 from dustanddreams/all_bugs_are_on_line_42

[bytecode] Make a cryptic comment slightly more helpful

shindere committed:
Merge pull request #13482 from shindere/ocamltest-fix-skipped-test-reporting

ocamltest should report skipped test as skipped rather than passed

dra27 committed:
Merge pull request #12397 from dra27/tweak-test

Tweak the dropped runtime events test

dra27 committed:
Merge pull request #13479 from shym/escape-cr

Escape `'\r'` in `sak`

gasche committed:
ocamltest_stdlib.ml: concatmap => List.concat_map

Suggested-by: Antonin Décimo <antonin@tarides.com>

dra27 committed:
Missing configure flag for BasicCompiler on tags

If the Hygiene workflow is triggered for a release tag (or for a branch
sat on a release tag) then ocamltest is not built by default, which
causes the .depend file to change. Explicitly enable ocamltest, even
though it is not built in BasicCompiler, so that the dependency
generation is consistent.

kit-ty-kate committed:
Add Format_doc.deprecated{,1} to ease transformations from Format to Format_doc by users of compiler-libs

kit-ty-kate committed:
Simplify some printer functions by using Format_doc.compat1

MisterDA committed:
Document new annotation macros

gasche committed:
Merge pull request #13340 from nojb/type_directed_array_literals

[RFC] Type-directed array literal disambiguation

gasche committed:
Revert "Merge pull request #13482 from shindere/ocamltest-fix-skipped-test-reporting"

This reverts commit 4689bc11224e8bfc346209021766cec828b1481e, reversing
changes made to a3c0e0f7690abe909c8ffce7716cfca92351c769.

Ahhj93 committed:
docs(introduction.md): fix link

gasche committed:
Merge pull request #13489 from Ahhj93/patch-1

docs(introduction.md): fix link

gasche committed:
Merge pull request #13491 from dra27/bang-bang

Remove stdlib/hashbang

tmcgilchrist committed:
Add changelog entry

Octachron committed:
Merge pull request #13487 from kit-ty-kate/format_doc_deprecated_printer1

Add Format_doc.deprecated{,1} to ease transformations from Format to Format_doc by users of compiler-libs

stedolan committed:
Merge pull request #13492 from dra27/ocamlrun-config-env

Include `CAML_LD_LIBRARY_PATH` in `ocamlrun -config`

dra27 committed:
Update Changes

Ahhj93 committed:
docs(asmcomp/riscv/NOTES.md): update deadlink (#13505)

gasche committed:
Merge branch 'trunk' into caml_stat-annotations

gasche committed:
Merge pull request #13457 from MisterDA/caml_stat-annotations

Annotate alloc/free open/close functions with compiler attributes

gasche committed:
Merge pull request #13316 from gasche/refactor-ocamltest-run

ocamltest: refactor `run_test_tree`

Octachron committed:
Revert "Merge pull request #13457 from MisterDA/caml_stat-annotations"

This reverts commit 2f5e828f842e04a00711bc9b115a0fa94aade011, reversing
changes made to 51a1781be2730f99ca33e4351d0cc5557b4c43cb.

dra27 committed:
Merge pull request #13468 from jmid/ocamlrunparam-manual-nit

Remove unused OCAMLRUNPARAMs list from manual

Octachron committed:
release howto: update to godbolt Compiler Explorer

gasche committed:
Merge pull request #13493 from dra27/force-linking

Improve linking error for force-custom libraries
